### PR TITLE
INF-718 - add support for rabbit mq

### PR DESCRIPTION
**Pull request by jamiegs**
Adds basic support for Rabbit-MQ.

### CODE REVIEW

File: tyr/servers/rabbit/server.py [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+    def __init__(self, group=None, server_type=None, instance_type=None,
+                    environment=None, ami=None, region=None, role=None,
+                    keypair=None, availability_zone=None, security_groups=None,
+                    block_devices=None, chef_path=None, subnet_id=None,
+                    dns_zones=None):
+
+        if server_type is None: server_type = self.SERVER_TYPE
+
+        super(RabbitMQServer, self).__init__(group, server_type, instance_type,
+                                environment, ami, region, role,
```
Comment (citruspi) [inline tyr/servers/rabbit/server.py]: These should be indented a little further.

File: tyr/servers/rabbit/server.py [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+                                    keypair, availability_zone,
+                                    security_groups, block_devices,
+                                    chef_path, subnet_id, dns_zones)
+
+    def configure(self):
```
Comment (citruspi) [inline tyr/servers/rabbit/server.py]: If you're not changing the behavior of the method - if you're just calling `super(...).configure()`, you don't need to re-define it.

File: tyr/servers/rabbit/server.py [hunk 3]
```diff
@@ -0,0 +1,35 @@
+from tyr.servers.server import Server
+import chef
```
Comment (citruspi) [inline tyr/servers/rabbit/server.py]: A lot of these imports don't seem to be used.

File: README.md [hunk 4]
```diff
@@ -42,6 +42,23 @@ node = MongoDataNode(group='monolith')
 node.autorun()
 ```
 
+## Examples:
+
+### Base Server
+``` python
+from tyr.servers.server import Server
```
Comment (citruspi) [inline README.md]: I'd recommend against advising users to do

```python
Server(...).autorun()
```

After the server finishes spinning up, there's no way to interact with it. Or if there's an issue, there's no way to rectify it. Or if the user wants to change a value, they will be unable to do that. I'd recommend doing something like

```python
s = Server(...)
s.autorun()
```

Then if it fails to save Chef attributes or the runlist and the problem is your network connection, you can just do

```python
s.bake()
```

If you'd like to get the server's hostname, you can just do

```python
s.hostname
```

If you'd like to check if `chef-client` has finished running, you can do

```python
s.baked()
```

If you'd like to run a command on the server, you can do

```python
s.run('sudo service mongod stop')
```

My point is that by storing the instance in a variable, you can continue to interact with it after running `autorun()`.

Comment (jamiegs) [inline README.md]: i've never had to interact with it after and I doubt most people would. I like the simplicity of just the two lines. I feel that Developers would what's going on with that snipping and know that you could store it to a variable if needed.

File: tyr/servers/rabbit/server.py [hunk 5]
```diff
@@ -0,0 +1,28 @@
...
+    CHEF_RUNLIST = ['role[RoleRabbit]']
+
+    IAM_ROLE_POLICIES = [
+        'allow-describe-tags',
+        'allow-volume-control',
+        'allow-update-route53-{environment}'
+    ]
+
+    def __init__(self, group=None, server_type=None, instance_type=None,
+                    environment=None, ami=None, region=None, role=None,
```
Comment (citruspi) [inline tyr/servers/rabbit/server.py]: These are indented a little too far - they should align with `self`.

**Comment by Z3r0Sum** [comment]
@citruspi @jamiegs This change is in conjunction with allowing dynamic values to be passed in for the EBS volume's IOPS and Size.  I've updated the cookbook as well to have the volume get created based on these dynamic attributes.

**Comment by jamiegs** [comment]
@Z3r0Sum these changes look good to me.

**Comment by Z3r0Sum** [comment]
@jamiegs  @citruspi Added a check for the IOPS/Size ratio of the EBS volume.

File: tyr/servers/rabbit/server.py [hunk 6]
```diff
@@ -0,0 +1,86 @@
...
+                    keypair=None, availability_zone=None, security_groups=None,
+                    block_devices=None, chef_path=None, subnet_id=None,
+                    dns_zones=None,vol_iops=500,vol_size=100,rabbit_user=None,rabbit_pass=None):
```
Comment (citruspi) [inline tyr/servers/rabbit/server.py]: This is over 79 characters.

Comment (citruspi) [inline tyr/servers/rabbit/server.py]: There should be spaces after the commas.

File: tyr/servers/rabbit/server.py [hunk 7]
```diff
@@ -0,0 +1,86 @@
...
+                                             security_groups, block_devices,
+                                             chef_path, subnet_id, dns_zones)
+
+    def configure(self):
+        """Make sure the IOPS to Size ratio is not greater than 30 for an EBS"""
+
+        super(RabbitMQServer, self).configure()
+
+        iops_size_ratio = self.vol_iops/self.vol_size
+        self.log.info('The IOPS to Size ratio is "{ratio}"'.format(ratio=iops_size_ratio))
```
Comment (citruspi) [inline tyr/servers/rabbit/server.py]: This is over 79 characters.

File: tyr/servers/rabbit/server.py [hunk 8]
```diff
@@ -0,0 +1,86 @@
...
+            self.rabbit_pass = rabbit_pass
+
+        super(RabbitMQServer, self).__init__(group, server_type, instance_type,
+                                             environment, ami, region, role,
+                                             keypair, availability_zone,
+                                             security_groups, block_devices,
+                                             chef_path, subnet_id, dns_zones)
+
+    def configure(self):
+        """Make sure the IOPS to Size ratio is not greater than 30 for an EBS"""
```
Comment (citruspi) [inline tyr/servers/rabbit/server.py]: This is over 72 characters.

File: tyr/servers/rabbit/server.py [hunk 9]
```diff
@@ -0,0 +1,86 @@
...
+
+        super(RabbitMQServer, self).configure()
+
+        iops_size_ratio = self.vol_iops/self.vol_size
+        self.log.info('The IOPS to Size ratio is "{ratio}"'.format(ratio=iops_size_ratio))
+        if iops_size_ratio > 30:
+            self.log.critical('The IOPS to Size ratio is greater than 30')
+            sys.exit(1)
+
+
```
Comment (citruspi) [inline tyr/servers/rabbit/server.py]: Extra blank line.

File: tyr/servers/rabbit/server.py [hunk 10]
```diff
@@ -0,0 +1,86 @@
...
+    def bake(self):
+        """Add IOPS and Volume Size attributes based on dynamic values from tyr"""
```
Comment (citruspi) [inline tyr/servers/rabbit/server.py]: This is over 72 characters.

File: tyr/servers/rabbit/server.py [hunk 11]
```diff
@@ -0,0 +1,86 @@
...
+            self.log.info('Set the rabbitmq user to '
+                          '{rabbit_user}'.format(rabbit_user=self.rabbit_user))
+
+            self.chef_node.attributes.set_dotted('rabbitmq.passwd',
+                                                 self.rabbit_pass)
+            self.log.info('Set the rabbitmq password to ... something secret')
+
+            self.chef_node.save()
+            self.log.info('Saved the Chef node configuration')
+
```
Comment (citruspi) [inline tyr/servers/rabbit/server.py]: Blank line.

File: tyr/servers/rabbit/server.py (hunk 8 above)
Comment (citruspi) [inline tyr/servers/rabbit/server.py]: Docstrings should be formatted

```
"""
Make sure the...
"""
```

File: tyr/servers/rabbit/server.py (hunk 10 above)
Comment (citruspi) [inline tyr/servers/rabbit/server.py]: Docstrings should be formatted

```
"""
Add IOPS and...
"""
```

**Pull request merged**